src/path_manager.c

```diff
@@ -831,14 +843,20 @@ struct mptcpd_pm *mptcpd_pm_create(struct mptcpd_config const *config)
                 return NULL;
         }
 
-        if (l_genl_add_family_watch(pm->genl,
-                                    MPTCP_GENL_NAME,
-                                    family_appeared,
-                                    family_vanished,
-                                    pm,
-                                    NULL) == 0) {
+        if (!l_genl_request_family(pm->genl,
+                                   MPTCP_GENL_NAME,
+                                   family_appeared,
+                                   pm,
+                                   NULL)
+            || l_genl_add_family_watch(pm->genl,
```

Comment:
This could race if the family appears after the request but before the watch. Is there any harm in creating a watch first and then requesting while the watch is active?

Reply:
That's a good point.  I believe that'll work after modifying the `family_appeared()` callback to return right away if it detects a previous call, e.g.:

```diff
diff --git a/src/path_manager.c b/src/path_manager.c
index 7676636..e4aefce 100644
--- a/src/path_manager.c
+++ b/src/path_manager.c
@@ -757,7 +757,17 @@ static void family_appeared(struct l_genl_family_info const *info,
 
         l_debug(MPTCP_GENL_NAME " generic netlink family appeared");
 
-        assert(pm->family == NULL);
+        /*
+          This function could be called in either of two cases, (1)
+          handling the appearance of the "mptcp" generic netlink
+          family through a family watch, or (2) through an explicit
+          family request.
+
+          Do nothing if the necessary "mptcp" family registration was
+          completed as a result of a previous call to this function.
+         */
+        if (pm->family != NULL)
+                return;
 
         pm->family = l_genl_family_new(pm->genl, MPTCP_GENL_NAME);
 
@@ -843,21 +853,21 @@ struct mptcpd_pm *mptcpd_pm_create(struct mptcpd_config const *config)
                 return NULL;
         }
 
-        if (!l_genl_request_family(pm->genl,
-                                   MPTCP_GENL_NAME,
-                                   family_appeared,
-                                   pm,
-                                   NULL)
-            || l_genl_add_family_watch(pm->genl,
+        if (l_genl_add_family_watch(pm->genl,
                                        MPTCP_GENL_NAME,
                                        family_appeared,
                                        family_vanished,
                                        pm,
-                                       NULL) == 0) {
+                                       NULL) == 0
+            || !l_genl_request_family(pm->genl,
+                                      MPTCP_GENL_NAME,
+                                      family_appeared,
+                                      pm,
+                                      NULL)) {
                 mptcpd_pm_destroy(pm);
-                l_error("Unable to request or watch \""
+                l_error("Unable to watch or request \""
                         MPTCP_GENL_NAME
-                        "\" Generic Netlink family.");
+                        "\" generic netlink family.");
                 return NULL;
         }
 
diff --git a/tests/test-commands.c b/tests/test-commands.c
index 9331252..9e9fb1a 100644
--- a/tests/test-commands.c
+++ b/tests/test-commands.c
@@ -181,13 +181,6 @@ int main(void)
         struct l_genl *const genl = l_genl_new();
         assert(genl != NULL);
 
-        bool const requested = l_genl_request_family(genl,
-                                                     MPTCP_GENL_NAME,
-                                                     run_tests,
-                                                     &tests_called,
-                                                     NULL);
-        assert(requested);
-
         unsigned int const watch_id =
                 l_genl_add_family_watch(genl,
                                         MPTCP_GENL_NAME,
@@ -198,6 +191,13 @@ int main(void)
 
         assert(watch_id != 0);
 
+        bool const requested = l_genl_request_family(genl,
+                                                     MPTCP_GENL_NAME,
+                                                     run_tests,
+                                                     &tests_called,
+                                                     NULL);
+        assert(requested);
+
         // Bound the time we wait for the tests to run.
         static unsigned long const milliseconds = 500;
         struct l_timeout *const timeout =
diff --git a/tests/test-path-manager.c b/tests/test-path-manager.c
index 523d078..0ea8e27 100644
--- a/tests/test-path-manager.c
+++ b/tests/test-path-manager.c
@@ -137,13 +137,6 @@ int main(void)
         struct l_genl *const genl = l_genl_new();
         assert(genl != NULL);
 
-        bool const requested = l_genl_request_family(genl,
-                                                     MPTCP_GENL_NAME,
-                                                     run_tests,
-                                                     &info,
-                                                     NULL);
-        assert(requested);
-
         unsigned int const watch_id =
                 l_genl_add_family_watch(genl,
                                         MPTCP_GENL_NAME,
@@ -154,6 +147,13 @@ int main(void)
 
         assert(watch_id != 0);
 
+        bool const requested = l_genl_request_family(genl,
+                                                     MPTCP_GENL_NAME,
+                                                     run_tests,
+                                                     &info,
+                                                     NULL);
+        assert(requested);
+
         // Bound the time we wait for the tests to run.
         static unsigned long const milliseconds = 500;
         struct l_timeout *const timeout =
```
I'll push this change.